lib/charms/haproxy/v0/haproxy_route_tcp.py

```diff
@@ -0,0 +1,1354 @@
+            )
+        except DataValidationError:
+            logger.error("Invalid requirer application data for %s", relation.app.name)
+            raise
+
+    def publish_proxied_endpoints(self, endpoints: list[str], relation: Relation) -> None:
+        """Publish to the app databag the proxied endpoints.
+
+        Args:
+            endpoints: The list of proxied endpoints to publish.
+            relation: The relation with the requirer application.
+        """
+        HaproxyRouteTcpProviderAppData(
+            endpoints=list(map(lambda x: cast(AnyHttpUrl, x), endpoints))
+        ).dump(relation.data[self.charm.app], clear=True)
+
+
+class HaproxyRouteTcpEnpointsReadyEvent(EventBase):
+    """HaproxyRouteTcpEnpointsReadyEvent custom event."""
+
+
+class HaproxyRouteTcpEndpointsRemovedEvent(EventBase):
+    """HaproxyRouteTcpEndpointsRemovedEvent custom event."""
+
+
+class HaproxyRouteTcpRequirerEvents(CharmEvents):
+    """List of events that the TLS Certificates requirer charm can leverage.
+
+    Attributes:
+        ready: when the provider proxied endpoints are ready.
+        removed: when the provider
+    """
+
+    ready = EventSource(HaproxyRouteTcpEnpointsReadyEvent)
+    removed = EventSource(HaproxyRouteTcpEndpointsRemovedEvent)
+
+
+class HaproxyRouteTcpRequirer(Object):
+    """haproxy-route interface requirer implementation.
+
+    Attributes:
+        on: Custom events of the requirer.
+    """
+
+    on = HaproxyRouteTcpRequirerEvents()
+
+    # pylint: disable=too-many-arguments,too-many-positional-arguments,too-many-locals
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        port: Optional[int] = None,
+        backend_port: Optional[int] = None,
+        hosts: Optional[list[str]] = None,
+        sni: Optional[str] = None,
+        check_interval: Optional[int] = None,
+        check_rise: Optional[int] = None,
+        check_fall: Optional[int] = None,
+        check_type: Optional[TCPHealthCheckType] = None,
+        check_send: Optional[str] = None,
+        check_expect: Optional[str] = None,
+        check_db_user: Optional[str] = None,
+        load_balancing_algorithm: Optional[LoadBalancingAlgorithm] = None,
+        load_balancing_consistent_hashing: bool = False,
+        rate_limit_connections_per_minute: Optional[int] = None,
+        rate_limit_policy: RateLimitPolicy = RateLimitPolicy.DENY,
+        upload_limit: Optional[int] = None,
+        download_limit: Optional[int] = None,
+        retry_count: Optional[int] = None,
+        retry_redispatch: bool = False,
+        server_timeout: Optional[int] = None,
+        connect_timeout: Optional[int] = None,
+        queue_timeout: Optional[int] = None,
+        server_maxconn: Optional[int] = None,
+        ip_deny_list: Optional[list[IPvAnyAddress]] = None,
+        enforce_tls: bool = True,
+        tls_terminate: bool = True,
+        unit_address: Optional[str] = None,
+    ) -> None:
```

Comment:
Add a guardrail to make sure most of the attributes are keywords args and not positional

Reply:
updated